polymorphia-frontend/interfaces/api/user/index.ts

```diff
@@ -1,10 +1,81 @@
 //TODO: handle which one is optional
-export interface UserDetailsDTO {
+export const Roles = {
+  STUDENT: "STUDENT",
+  INSTRUCTOR: "INSTRUCTOR",
+  COORDINATOR: "COORDINATOR",
+  UNDEFINED: "UNDEFINED",
+} as const;
+
+export type Role = (typeof Roles)[keyof typeof Roles];
+
+export const RoleTextMap: Record<Role, string> = {
+  [Roles.STUDENT]: "Student",
+  [Roles.INSTRUCTOR]: "Prowadzący",
+  [Roles.COORDINATOR]: "Koordynator",
+  [Roles.UNDEFINED]: "Nieokreślony",
+};
+
+export interface BaseUserDetails {
   id: number;
-  studentName: string;
+  userName: string;
+  courseId: number;
+  imageUrl: string;
+}
+
+interface BaseUserDetailsDTOWithType<
+  T extends Role,
+  R extends BaseUserDetails,
+> {
+  userType: T;
+  userDetails: R;
+}
+
+export interface StudentDetailsDTO extends BaseUserDetails {
   animalName: string;
   evolutionStage: string;
   group: string;
-  imageUrl: string;
   position: number;
 }
+
+export type StudentDetailsDTOWithType = BaseUserDetailsDTOWithType<
+  "STUDENT",
+  StudentDetailsDTO
+>;
```

Comment:
```suggestion
export type StudentDetailsDTOWithType = BaseUserDetailsDTOWithType<
  typeof Roles.STUDENT,
  StudentDetailsDTO
>;
```

applies to all below.